### DISCUSSION TITLE
Getting NVENC to work right

### DISCUSSION THREAD

**Discussion by TheSashmo** [CESNET/UltraGrid#329] [General]
Sorry for the millions of posts in the last few weeks.  I am trying UG with nvenc, and it seems that everything I do results back to failure.
- Cuda drivers installed (validated with nvidia-smi) NVIDIA-SMI 470.182.03   Driver Version: 530.41.03    CUDA Version: 12.1
- Compiled UG with cuda support:
- UltraGrid configuration summary
===============================
Target: x86_64-pc-linux-gnu

Features:
  Crypto (OpenSSL) ................ yes   CUDA support (nvcc, default) .... yes
  Debug output .................... no    iHDTV support ................... no
  IPv6 support .................... yes   Library live555 ................. no
  OpenCV .......................... yes   Profiling support ............... no
  Qt GUI .......................... no    RT priority ..................... no
  SpeexDSP ........................ no    Soxr ............................ no
  Standalone modules .............. no    zfec ............................ yes

Audio:
  ALSA ............................ yes   CoreAudio ....................... no
  JACK ............................ no    JACK transport .................. no
  SDL_mixer ....................... no    Portaudio ....................... no
  WASAPI .......................... no

Video:
  AJA ............................. no    AV Foundation ................... no
  BitFlow ......................... no    Bluefish444 ..................... no
  caca ............................ no    DeckLink ........................ yes
  DELTACAST ....................... no    DirectShow ...................... no
  DVS ............................. no    File input ...................... yes
  Gpustitch ....................... no    NDI ............................. yes
  OpenGL .......................... no    OpenXR VR Display ............... no
  Panorama Gl Display ............. no    RTSP capture client ............. yes
  SAGE ............................ no    Screen capture (X11) ............ yes
  SDL ............................. no    SW video mix .................... no
  V4L2 ............................ yes   VULKAN_SDL2 ..................... no
  XIMEA ........................... no

Compressions:
  Cineform ........................ no    Comprimato J2K .................. no
  CUDA DXT ........................ yes   GPUJPEG ......................... yes
  GPUJPEG transcode to DXT ........ no    Lavc (VDP yes, VA yes, RPI4 no) . yes
  Realtime DXT .................... no    UYVY dummy compression .......... no

Others:
  Blank capture filter ............ yes   GPU accelerated LDGM ............ yes
  Hole punching ................... no    iHDTV support ................... no
  MCU-like video mixer ............ yes   NAT-PMP traversal ............... no
  PCP NAT traversal ............... no    Resize capture filter ........... yes
  RTSP server ..................... no    Scale postprocessor ............. no
  SDP over HTTP ................... no    Spout ........................... no
  Syphon .......................... no    Testcard extras ................. yes
  Text postprocess and filter ..... no    Swscale support ................. yes

*** Following WARNINGS were issued:
Neither Soxr nor SpeexDSP was not found. Strongly recommending installing that, otherwise audio part of UG will be crippled.

checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/config.h
config.status: executing depfiles commands

Testing this command:
`./uv -t decklink:4 -s embedded -c libavcodec:encoder=nvenc_hevc:bitrate=20000k --audio-codec=MP3:sample_rate=48000:bitrate=256k --audio-capture-format channels=16 -m 1316 -P 10000 192.168.99.187 --control-port 12345`

And I get: 
[lavc nvenc_hevc @ 0x7fea880979c0] dl_fn->cuda_dl->cuInit(0) failed -> CUDA_ERROR_UNKNOWN: unknown error
[lavc] Could not open codec for pixel format rgb0

Tried the same with test card as a source:
[lavc nvenc_hevc @ 0x7f29cc0f13c0] This encoder is deprecated, use 'hevc_nvenc' instead
[lavc nvenc_hevc @ 0x7f29cc0f13c0] dl_fn->cuda_dl->cuInit(0) failed -> CUDA_ERROR_UNKNOWN: unknown error
[lavc] Could not open codec for pixel format yuv444p16le

I played around with other pixel formats:
[lavc nvenc_hevc @ 0x7f29cc0f3280] This encoder is deprecated, use 'hevc_nvenc' instead
[lavc nvenc_hevc @ 0x7f29cc0f3280] dl_fn->cuda_dl->cuInit(0) failed -> CUDA_ERROR_UNKNOWN: unknown error
[lavc] Could not open codec for pixel format rgb0

Suggestions?

**Comment by alatteri** [comment]
try the AppImage and encoder= hevc_nvenc

**Comment by TheSashmo** [comment]
Thanks but, Same result.

**Comment by alatteri** [comment]
What OS? What Nvidia card? What is the SDI source format?  What install steps for Nvidia driver? 
Have you tried with testcard?

**Comment by TheSashmo** [comment]
Yes with testcard.
Ubuntu 20.04
Nvidia Quadro P600
Tried varying SDI inputs, normally on auto detect, but tried 1080 all the way to 4k.

**Comment by MartinPulec** [comment]
`ffmpeg -f lavfi -i testsrc -c:v hevc_nvenc -t 1 out.mp4` works?

**Comment by TheSashmo** [comment]
Same error:
Input #0, lavfi, from 'testsrc':
  Duration: N/A, start: 0.000000, bitrate: N/A
    Stream #0:0: Video: rawvideo (RGB[24] / 0x18424752), rgb24, 320x240 [SAR 1:1 DAR 4:3], 25 tbr, 25 tbn, 25 tbc
Stream mapping:
  Stream #0:0 -> #0:0 (rawvideo (native) -> hevc (hevc_nvenc))
Press [q] to stop, [?] for help
[hevc_nvenc @ 0x55818763b140] dl_fn->cuda_dl->cuInit(0) failed -> CUDA_ERROR_UNKNOWN: unknown error
Error initializing output stream 0:0 -- Error while opening encoder for output stream #0:0 - maybe incorrect parameters such as bit_rate, rate, width or height
Conversion failed!

**Comment by MartinPulec** [comment]
okay, so if it is not UG issue, I'll convert it rather to discussion.

**Reply by TheSashmo** [reply]
looks like its a driver vs. kernel issue....

**Reply by MartinPulec** [reply]
I am interested in this:

> NVIDIA-SMI 470.182.03 Driver Version: 530.41.03

do you use distro packages or driver directly downloaded from NVIDIA website? I'd perhaps get rid of nvidia-*-470 altogether, eg.:

`sudo aptitude remove "~i?name(nvidia.*470)"`

just to ensure that there isn't mixinig something together

**Reply by MartinPulec** [reply]
nvidia-smi 530 should be in a package _nvidia-utils-530_, I believe. _nvidia-smi_ doesn't complain on anything, anyways? (I mean the 470 one, but also 530 would be useful  to check).

**Reply by alatteri** [reply]
I don't use NVenc, but I found this in my notes.

`sudo apt-get -y install libffmpeg-nvenc-dev`

**Reply by MartinPulec** [reply]
this is not needed, you'd need this if you compile FFmpeg by yourself with NVENC support. But if you have FFmpeg compiled with that support (which already is packaged in U20.04), this shouldn't be necessary. AFAIK only the actual driver is needed and I believe that also libnvidia-compute-XXX (version corresponding the driver), but this should be handled by dependencies (if you have the _nvidia-driver-XXX_ package installed).

**Comment by TheSashmo** [comment]
Alright, we can close this up.... So much headache for something that should be much easier.  I originally installed the drivers from apt-get, and clearly that didnt work.  I had to spend an hour purging and eventually just went to the nvidia website and downloaded the correct version.  The issue is right in the top line of nvidia smi.

This is correct:
 NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2

This is incorrect:
NVIDIA-SMI 470.182.03              Driver Version: 530.41.03   CUDA Version: 12.1

**Comment by TheSashmo** [comment]
Why we are still here.... 

[lavc] Average compression time of last 100 frames is 22.909312 ms but time per frame is only 20.000000 ms!
[lavc] Consider adding "delay=<frames>" option to NVENC compression (2 suggested) to increase throughput at the expense of latency

**Reply by MartinPulec** [reply]
Interesting, do you have full (expected) FPS in this case? This message was added as a safety check to guard if a compression performance is sufficient. For that it measures the duration of compression and compares it with frame time (1.0/FPS).

From your data it is a 50p signal... I believe that the card should be able to compress at least at 4K without a problem. But again, it depends on context - if it occurs eg. only once at the beginning and everything is working correctly, I woudn't care at all, it is just hint. If the compression is indeed not performing well, you can try the hint as it is suggesting. Feel free to open a new issue - I believe that NVENC should not have such performance problems, but I don't know details.

If "delay=1" or "delay=2" helps, let me know. Default is currently 0, but 1 may be a good performance/latency compromise.

**Reply by MartinPulec** [reply]
I've tested on a _3080_ (but it will likely to be valid for any of Ampre-line card) with `-t testcard:size=3840x2160:pattern=noise` and (as seen with `-VV`), the compression took 22 ms with _delay_ 0 (implicit) and 1 and _7_ ms with delay 2. It seems like the numbers are valid regardless the video content, maybe _2_ frames could be the default value but since there is the hint, I don't think there is an issue
